File: modules/expo-godot-view/ios/ExpoGodotViewModule.swift
import ExpoModulesCore

public class ExpoGodotViewModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoGodotView")

    // Godot → host. Carries an opaque JSON envelope string in `message`.
    Events("onGodotMessage")

    OnCreate {
      GodotBridge.shared.onMessageToHost = { [weak self] json in
        self?.sendEvent("onGodotMessage", ["message": json])
      }
    }

    OnDestroy {
      GodotBridge.shared.onMessageToHost = nil
    }

    // Host → Godot. `message` is a JSON envelope string `{ "type": ..., "data": ... }`.
    // No-op on the simulator (no engine).
    Function("sendMessageToGodot") { (message: String) in
      GodotBridge.shared.addMessage(message)
    }

    View(ExpoGodotView.self) {
      Prop("sceneName") { (view: ExpoGodotView, name: String?) in
        view.sceneName = name
      }
      // True only when the host wants drag-to-orbit on this view (currently the Controls screen).
      // Off elsewhere so GLView's touchesBegan stays blocked (root-cause workaround).
      Prop("orbitEnabled") { (view: ExpoGodotView, enabled: Bool?) in
        view.orbitEnabled = enabled ?? false
      }
    }
  }
}

File: modules/expo-godot-view/android/src/main/java/expo/modules/godotview/ExpoGodotViewModule.kt
package expo.modules.godotview

import androidx.core.os.bundleOf
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * JS <-> Godot bridge. Mirrors `ios/ExpoGodotViewModule.swift`; the JS contract
 * (`sendMessageToGodot` + the `onGodotMessage` event) is identical on both platforms.
 */
class ExpoGodotViewModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoGodotView")

    // Godot → host. Carries an opaque JSON envelope string in `message`.
    Events("onGodotMessage")

    OnCreate {
      GodotBridge.onMessageToHost = { json -> sendEvent("onGodotMessage", bundleOf("message" to json)) }
    }

    OnDestroy {
      GodotBridge.onMessageToHost = null
      GodotBridge.clear()
    }

    // Host → Godot. `message` is a JSON envelope string `{ "type": ..., "data": ... }`, drained by
    // MobileComm.gd through AndroidGodotInterface.pendingMessagesCount()/getMessage().
    Function("sendMessageToGodot") { message: String ->
      GodotBridge.addMessage(message)
    }

    View(ExpoGodotView::class) {
      Prop("sceneName") { view: ExpoGodotView, name: String? ->
        view.sceneName = name
      }
    }
  }
}
